src/chunk.js

```diff
@@ -99,53 +115,258 @@ export function runChunksForTesting(nodeOrAmpDoc) {
   }
 }
 
+/**
+ * The priority of a chunk task. Higher priority tasks have higher values.
+ * @enum {number}
+ */
+export const ChunkPriority = {
+  HIGH: 20,
+  LOW: 10,
+  BACKGROUND: 0,
+};
+
+/** @enum {string} */
+const TaskState = {
+  NOT_RUN: 'not_run',
+  RUN: 'run',
+};
+
+/**
+ * A default chunkable task.
+ */
+class Task {
+  /**
+   * @param {!Function} fn
+   */
+  constructor(fn) {
+    /** @public {TaskState} */
+    this.state = TaskState.NOT_RUN;
+
+    /** @private @const {!Function} */
+    this.fn_ = fn;
+
+    /** @private {?Function} */
+    this.resolve_ = null;
+
+    /** @private {?Function} */
+    this.reject_ = null;
+
+    /** @public @const */
+    this.promise = new Promise((resolve, reject) => {
+      this.resolve_ = resolve;
+      this.reject_ = reject;
+    });
+  }
+
+  /**
+   * Executes the wrapped function.
+   * @throws {Error}
+   */
+  runTask() {
+    this.state = TaskState.RUN;
+    try {
+      const result = this.fn_();
+      this.resolve_(result);
+    } catch (e) {
+      this.onTaskError_(e);
+      this.reject_(e);
+      throw e;
+    }
+  }
+
+  /**
+   * @return {string}
+   */
+  get name() {
```

Comment:
I think we decided not to use getters like this since they are somewhat slow. /cc @jridgewell . 

Also, `displayName` is good for local debug, but most are likely stripped by compiler for prod binaries.